File: leetcode/easy/count-equal-and-divisible-pairs-in-an-array.ts
// https://leetcode.com/problems/count-equal-and-divisible-pairs-in-an-array/

// MINE: brute force
/*
function countPairs(nums: number[], k: number): number {
    let ans = 0
    for(let i=0; i<nums.length; ++i) {
        for(let j=i+1; j<nums.length; ++j) if(nums[i] === nums[j] && (i*j) % k === 0) ++ans
    }
    return ans
};
*/

// ANSWER
function countPairs(nums: number[], k: number): number {
  const freq = new Map()
  let ans = 0
  for (let i = 0; i < nums.length; ++i) {
    const vals = freq.get(nums[i])
    vals ? freq.set(nums[i], [...vals, i]) : freq.set(nums[i], [i])
  }
  for (let vals of freq.values()) {
    const gcds = new Map()
    for (let i of vals) {
      let gcdI = gcd2(i, k)
      for (let [gcdJ, count] of gcds) if ((gcdI * gcdJ) % k === 0) ans += count
      gcds.set(gcdI, gcds.get(gcdI) + 1 || 1)
    }
  }
  return ans
}

const gcd2 = function (a: number, b: number): number {
  if (b === 0) return a
  return gcd(b, a % b)
}

File: leetcode/easy/count-equal-and-divisible-pairs-in-an-array.py
# https://leetcode.com/problems/count-equal-and-divisible-pairs-in-an-array/

class Solution:
    def countPairs(self, nums: List[int], k: int) -> int:
        freq = {}
        co = 0
        for i in range(len(nums)):
            if nums[i] not in freq:
                freq[nums[i]] = []
            freq[nums[i]].append(i)
        for vals in freq.values():
            get_gcds = {}
            for i in vals:
                gcd_i = gcd(i, k)
                for gcd_j, cnt in get_gcds.items():
                    if gcd_i * gcd_j % k == 0:
                        co += cnt
                if gcd_i in get_gcds:
                    get_gcds[gcd_i] += 1
                else:
                    get_gcds[gcd_i] = 1
        return co
